### PR TITLE
[#605] refactor nanogpt dataloader to PyTorch standard

### PR DESCRIPTION
**Pull request by SylviaSyp**
### What changes were proposed in this pull request?

convert nanogpt dataloader to a formal torch style dataloader.

### Why are the changes needed?

this refactor aims to enhance code readability, make it more maintainable and modular, and facilitate better integration with the PyTorch ecosystem.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

by running the refactored code.

### CODE REVIEW

File: model_zoo/pytorch/nanogpt/train.py [hunk 1]
```diff
@@ -27,65 +27,50 @@
...
-    meta_vocab_size = meta["vocab_size"]
-    print(f"found vocab_size = {meta_vocab_size} (inside {meta_path})")
-    return train_data, val_data, meta_vocab_size
+class GPTDataset(Dataset):
+    def __init__(self, data_path, block_size=128):
+        self.data = np.memmap(data_path, dtype=np.uint16, mode="r")
+        self.block_size = block_size
 
+    def __len__(self):
+        return len(self.data) - self.block_size
```
Comment (workingloong) [inline model_zoo/pytorch/nanogpt/train.py:44]: Why do we need to minus the `self.block_size`?

File: model_zoo/pytorch/nanogpt/train.py [hunk 2]
```diff
@@ -27,65 +27,50 @@
...
+    val_dataset = GPTDataset(os.path.join(data_dir, "val.bin"), block_size)
+    with open(os.path.join(data_dir, "meta.pkl"), "rb") as f:
+        meta = pickle.load(f)
+    train_loader = DataLoader(
+        train_dataset, batch_size=batch_size, shuffle=True
     )
-    if device_type == "cuda":
-        # Pin arrays x,y, which allows us to move them
-        # to GPU asynchronously (non_blocking=True)
-        x, y = x.pin_memory().to(device, non_blocking=True), y.pin_memory().to(
```
Comment (workingloong) [inline model_zoo/pytorch/nanogpt/train.py:83]: Can use `pin_memory` in the dataloader?

Comment (SylviaSyp) [reply model_zoo/pytorch/nanogpt/train.py:83]: Yes, it was my oversight, and I will add this implementation.

**Review by workingloong — APPROVED** [review]
Perfect!

**Pull request merged**